docs/Network/RaspberryPi_Devices/LTE_CAT_4_EG25_GL_4G_HAT_for_Raspberry_Pi/Applications/getting_start_with_raspberry_pi_4g_hat_and_quectel_gnss.md

```diff
@@ -0,0 +1,283 @@
+            latitude_raw, longitude_raw = data[1], data[2]
+
+            # Correct parsing by removing N/S/E/W at the end
+            latitude = convert_to_decimal(latitude_raw[:-1], 'N' in latitude_raw)
+            longitude = convert_to_decimal(longitude_raw[:-1], 'E' in longitude_raw)
+
+            print(f"Latitude: {latitude}, Longitude: {longitude}")
+            return latitude, longitude
+        except (IndexError, ValueError) as e:
+            print(f"Error parsing coordinates: {e}")
+            return None, None
+    return None, None
+
+
+def convert_to_decimal(coord, positive=True):
+    """Convert NMEA-style coordinate to decimal format."""
+    # Split the coordinate into degrees and minutes
+    degrees = int(float(coord) // 100)
+    minutes = float(coord) % 100
+    decimal = degrees + (minutes / 60)
+
+    # If it's in the Southern or Western hemisphere, make it negative
+    return decimal if positive else -decimal
+
+
+def get_location(lat, lon):
+    """Get a human-readable location using geopy."""
+    geolocator = Nominatim(user_agent="gnss_locator")
+    try:
+        location = geolocator.reverse((lat, lon), exactly_one=True)
+        return location.address if location else "Location not found"
+    except Exception as e:
+        print(f"Geocoding error: {e}")
+        return "Error retrieving location"
+
+
+def automate_gnss(port, baudrate=9600):
+    """Automate GNSS commands and fetch location."""
+    try:
+        with serial.Serial(port, baudrate, timeout=1) as ser:
+            print("Serial port connected.")
+
+            # Turn on GNSS
+            send_at_command(ser, "AT+QGPS=1", delay=2)
+
+            # Get positioning information
+            response = send_at_command(ser, "AT+QGPSLOC=0", delay=2)
+            latitude, longitude = parse_coordinates(response)
+
+            if latitude and longitude:
+                print(f"Coordinates: Latitude = {latitude}, Longitude = {longitude}")
+                location = get_location(latitude, longitude)
+                print(f"Location: {location}")
+            else:
+                print("Unable to retrieve coordinates.")
+
+            # Turn off GNSS
+            send_at_command(ser, "AT+QGPSEND", delay=2)
+            print("GNSS process completed.")
+
+    except serial.SerialException as e:
+        print(f"Error: {e}")
+
+
+# Set your USB port (e.g., '/dev/ttyUSB0' or 'COM3' on Windows)
+usb_port = "/dev/ttyUSB2"  # Update based on your system
+automate_gnss(usb_port)
+
+
+```
+
+### Step 3. Run the Script
+
+- Open a terminal, ensure you're in the project directory:
+```bash
+cd GNSS_EX
+```
+
+- Activate the virtual environment:
+
+```bash
+source env/bin/activate
+```
+
+- Run the script using Python:
+```bash
+python test_gnss.py
+```
+- The output 
+  
+<div style={{textAlign:'center'}}><img src="https://files.seeedstudio.com/wiki/4g_hat_raspberry_pi_eg25_gl/gnss_output.PNG" style={{width:800}}/></div>
+
```

Comment:
The last section of the document seems to end abruptly. Could you please confirm if it has been completed?

Reply:
Sorry ! my fault now it is corrected !!! thank you for point out !!!!!!!